File: backend/RoomBooking.Api/Shared/Http/ApiResponse.cs
namespace RoomBooking.Api.Shared.Http;

public sealed class ApiResponse<T>
{
    public bool Success { get; init; }
    public T? Data { get; init; }
    public ApiError? Error { get; init; }

    public static ApiResponse<T> Ok(T? data = default) => new()
    {
        Success = true,
        Data = data
    };

    public static ApiResponse<T> Fail(string detail, string? code = null) => new()
    {
        Success = false,
        Error = new ApiError(detail, code)
    };
}

File: frontend/src/lib/api-response.ts
import axios from 'axios'

import type { ApiResponse } from '@/types/api'

export function unwrapApiResponse<T>(envelope: ApiResponse<T>): T {
  if (!envelope.success || envelope.data == null) {
    throw new Error(envelope.error?.detail ?? 'Request failed')
  }
  return envelope.data
}

export function getApiErrorMessage(
  error: unknown,
  fallback = 'Something went wrong. Please try again.',
): string {
  if (!axios.isAxiosError(error)) return fallback

  const data = error.response?.data as
    | { error?: { detail?: unknown }; detail?: unknown }
    | undefined

  const nested = data?.error?.detail
  if (typeof nested === 'string' && nested.trim()) return nested

  const legacy = data?.detail
  if (typeof legacy === 'string' && legacy.trim()) return legacy

  return fallback
}
